### ISSUE TITLE
머딥 - 22.1 (tensorflow 심화 + cnn)

### ISSUE THREAD

**Issue by vmtmxmf5** [vmtmxmf5/Python-Basics#61]
# 목차
1. tf.data
2. ds + Dropout 
3. Functional API 준비
3. Functional API 훈련

### Summary
| 용어 | 설명 |
|--|--|
| train_ds.take(1)  | 첫 번째 배치를 꺼내온다 |
| shuffle | 훈련용o, 테스트용x |
| Dropout | 레이어 사이사이에 넣어야 함 |
| @tf.function | tf.session에 커스텀 할 함수를 올리는 용도 |
| SparseCategoricalCrossentropy | ~OH metric |
| metrics.Mean() | 배치별 손실함수 평균 구할 때 사용하는 평가지표 |
| with tf.GradientTape() as tape | tf.GradientTape().open() <br> 추가할 내용 ~~ <br> tf.GradientTape().close() 와 같다 <br> 자동 미분 준비과정 |
| train loop 준비과정 | (1) tf.GradientTape()에서 pred, loss 준비 <br> (2) tape.gradient 그라디언트 벡터 계산 <br> (3) optm.apply_gradient(zip()) 업데이트

데이터 준비
```
import matplotlib.pyplot as plt
import tensorflow as tf
import numpy as np

from tensorflow.keras import datasets
mnist = datasets.mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

X_train = X_train[..., tf.newaxis] / 255.
X_test = X_test[..., tf.newaxis] / 255.

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import *
```

## TF data
배열을 **병렬**로 처리

- from_tensor_slices()
feature와 label을 하나로 합친 ds(dataset) 생성

- shuffle()
ds 데이터를 섞어줌

- batch()
배치 생성하기

```sage
train_ds = tf.data.Dataset.from_tensor_slices((X_train, y_train))
train_ds = train_ds.shuffle(1024) # 데이터를 1kb 단위로 섞는다 (임시공간 버퍼에서) (!= 랜덤시드)
train_ds = train_ds.batch(32)
```

#### 주의 ! **shuffle은 훈련용**
```sage
test_ds = tf.data.Dataset.from_tensor_slices((X_test, y_test))
test_ds = test_ds.batch(32) #test로 훈련하는게 아니기 때문에 굳이 shuffle 할 필요 x
```

- take()
이미 정해진 배치 사이즈만큼 데이터를 꺼내올 수 있다

```sage
# 첫번째 배치만 꺼내오겠다
for images, labels in train_ds.take(1):
    print(images.shape, labels.shape)
    image = images[0, ..., 0]
    label = labels.numpy()[0]
    
    plt.title(label)
    plt.imshow(image, cmap='gray')
plt.show()
```

## ds + Dropout
```
from tensorflow.keras.layers import Dropout
```

- Dropout
레이어 사이사이마다 넣어줘야 함

```sage
model = Sequential([
    Conv2D(32, 3, padding='SAME', input_shape=(28, 28, 1), activation='relu'),
    Conv2D(32, 3, padding='SAME', activation='relu'),
    MaxPool2D(pool_size=(2, 2)),
    
    Dropout(0.25),
    
    Conv2D(64, 3, padding='SAME', activation='relu'),
    Conv2D(64, 3, padding='SAME', activation='relu'),
    MaxPool2D(pool_size=(2,2)),
    
    Dropout(0.25),
    
    Flatten(),
    Dense(512, activation='relu'),
    Dense(10, activation='softmax')
    ])
model.summary()

model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['acc'])

# validation_data는 따로 만들어야 한다
model.fit(train_ds, epochs=1)
```

## Functional API 준비
tf.keras.layers 상속
loss, optm 직접 정의

- @tf.function 이란?
tensorflow 의 로직(계산그래프)에 커스텀 함수를 올린다.
즉, tf.session에서 관리하는 함수로 만든다는 뜻

```sage
loss_object = tf.keras.losses.SparseCategoricalCrossentropy()
optimizer = tf.keras.optimizers.Adam()

train_loss = tf.keras.metrics.Mean(name='train_loss') # 배치별 로스 평균
train_acc = tf.keras.metrics.SparseCategoricalAccuracy(name='train_acc') # ~OH accuracy

test_loss = tf.keras.metrics.Mean(name='test_loss')
test_acc = tf.keras.metrics.SparseCategoricalAccuracy(name='test_acc')
```

- with란?
tf.GradientTape().open()
사이에서 함수 변경 가능
tf.GradientTape().close()

```sage
@tf.function
def train_step(images, labels):
    # 함수 안에 하이퍼파라미터 추가가능
    with tf.GradientTape() as tape:
    # 역전파를 이용한 자동 미분 시스템 구축
    # 미분을 수행하는 게 아니라 준비과정임 (주의)
        
        # 1. 예측
        prediction = model(images) #모델 훈련모드
                
        # 2. loss
        loss = loss_object(labels, prediction)
    
    # 3. 오차 역전파 수행 (그라디언트 벡터 구하기)
    grads = tape.gradient(loss, model.trainable_variables) # train_vari. = 가중치 편향
    
    # 4. 최적화(편미분 값 grads를 넣어야함)
    optimizer.apply_gradients(zip(grads, model.trainable_variables))
    ## apply_gradient를 통해 최적화 안에 변화량과 파라미터 넣어줘야 함
    
    train_loss(loss)
    train_acc(labels, prediction)
```
- 언제 train loop 쓸까?
전이학습과 콜라보 할 때 API 사용
혹은 내가 만든 loss나 optm 쓸 때 사용

```sage
# test세트는 역전파 할 필요가 없으므로!
@tf.function 
def test_step(images, labels):
    prediction = model(images)
    t_loss = loss_object(labels, prediction)
    
    test_loss(t_loss)
    test_acc(labels, prediction)
```

## Functional API 훈련 (compile + fit)
```sage
epochs = 10

for epoch in range(epochs):
    # take를 쓰지 않았으므로 모든 데이터 불러옴
    for images, labels in train_ds:
        train_step(images, labels)
    
    for test_images, test_labels in test_ds:
        test_step(test_images, test_labels)
    
    template = 'Epoch {}, Train_Loss {:.3f}, Train_acc{:.3f}, Test_Loss {:.3f}, Test_acc{:.3f}'
    print(template.format(epoch + 1, train_loss.result(), train_acc.result(), test_loss.result(), test_acc.result()))
```
